ttnn/cpp/ttnn/operations/data_movement/slice/slice.cpp

```diff
@@ -8,133 +8,239 @@
         auto& input_tensor_shape = input_tensor.get_legacy_shape();
+        auto memory_config = memory_config_arg.value_or(input_tensor.memory_config());
         if (input_tensor.is_sharded() && input_tensor.memory_config() == memory_config &&
-            input_tensor_shape.rank() > 1 && input_tensor_shape.rank() == output_tensor_start.rank() &&
-            output_tensor_start.rank() == output_tensor_end.rank()) {
-            TT_FATAL(!modified_step.has_value(), "Sharded tensor slice implementation does not support striding");
+            input_tensor_shape.rank() > 1) {
+            TT_FATAL(no_step, "Sharded tensor slice implementation does not support striding");
             uint32_t i;
             // Require all leading dims to be 1 (TODO: This can be relaxed to support outermost non-1 dim unpadding)
             bool in_place_unpad = true;
             for (i = 0; i < input_tensor.get_legacy_shape().rank() - 2; ++i) {
                 in_place_unpad &=
-                    output_tensor_start[i] == 0 && output_tensor_end[i] == 0 && input_tensor_shape[i] == 1;
+                    modified_begins[i] == 0 && modified_ends[i] == 1 && input_tensor_shape[i] == 1;
             }
-            in_place_unpad &= output_tensor_start[i] == 0 &&
-                              tt::div_up(output_tensor_end[i] + 1, input_tensor.shard_spec().value().shape[0]) ==
+            in_place_unpad &= modified_begins[i] == 0 &&
+                              tt::div_up(modified_ends[i], input_tensor.shard_spec().value().shape[0]) ==
                                   tt::div_up(input_tensor_shape[i], input_tensor.shard_spec().value().shape[0]);
             i++;
-            in_place_unpad &= output_tensor_start[i] == 0 && output_tensor_end[i] == input_tensor_shape[i] - 1;
+            in_place_unpad &= modified_begins[i] == 0 && modified_ends[i] == input_tensor_shape[i];
             if (in_place_unpad) {
                 auto new_shape = input_tensor.get_legacy_shape();
                 auto new_pad = new_shape.padding();
 
-                std::size_t unpad_val = input_tensor_shape[-2] - output_tensor_end[-2] - 1;
+                std::size_t unpad_val = input_tensor_shape[-2] - modified_ends[-2];
                 new_shape[-2] -= unpad_val;
                 new_pad[-2].back -= std::min(unpad_val, new_pad[-2].back);
                 auto padded_shape = ttnn::Shape(tt::tt_metal::LegacyShape(new_shape, new_pad));
                 return Tensor(input_tensor.storage(), padded_shape, input_tensor.dtype(), input_tensor.layout());
             }
         }
 
-        return operation::run(
-                   SliceDeviceOperation{output_tensor_start, output_tensor_end, modified_step, memory_config}, {input_tensor}, {}, {}, queue_id)
+        auto res = operation::run(
+                   SliceDeviceOperation{
+                    tt::tt_metal::LegacyShape(modified_begins),
+                    tt::tt_metal::LegacyShape(padded_ends),
```

Comment:
these really should be vectors, right?

Reply:
Arrays even as they're fixed to 4D I believe.